### PR TITLE
Create all punches endpoint

### PR DESCRIPTION
**Pull request by Thekote**
Issue: #104 

This PR adds an endpoint to list all punches registered to the authenticated user.

This is how the response is working:

```json
[
    {
        "created_at": "2022-04-26T09:53:51.248-03:00",
        "from": "2021-10-26T08:00:00.000-03:00",
        "to": "2021-10-26T12:00:00.000-03:00",
        "delta_as_hour": "04:00",
        "extra_hour": false,
        "project": {
            "name": "Rito Gomes"
        }
    },
    {
        "created_at": "2022-04-26T09:53:51.252-03:00",
        "from": "2021-10-26T13:00:00.000-03:00",
        "to": "2021-10-26T16:00:00.000-03:00",
        "delta_as_hour": "03:00",
        "extra_hour": false,
        "project": {
            "name": "Rito Gomes"
        }
    }
```

### CODE REVIEW

File: app/controllers/api/v1/punches_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    class PunchesController < ActionController::API
+      before_action :authenticate_user!
+
+      def index
+        render json: scopped_punches, status: :ok, each_serializer: PunchSerializer
+      end
+
+      private
+
+      def scopped_punches
```
Comment (rhian-cs) [inline app/controllers/api/v1/punches_controller.rb:12]:
```suggestion
      def scoped_punches
```
Are these supposed to be called `scoped_punches`?

File: spec/controllers/api/v1/punches_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+require 'spec_helper'
+
+describe Api::V1::PunchesController, :type => :controller do
+
+  let!(:user) { create(:user) }
+  let!(:project) { create(:project, company: user.company) }
+  let!(:company) { user.company }
+  let!(:punches) { Punch.create!(created_at: '2022-04-26T09:53:51.248-03:00', from: '2021-10-26T08:00:00.000-03:00', to: '2021-10-26T12:00:00.000-03:00', extra_hour: false, project_id: project.id, company_id: company.id, user_id: user.id) }
```
Comment (rhian-cs) [inline spec/controllers/api/v1/punches_controller_spec.rb:8]: This line seems a little long, you could perhaps split this into more lines.

File: spec/controllers/api/v1/punches_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  before { login user }
+
+  describe 'GET api/v1/punches' do
```
Comment (rhian-cs) [inline spec/controllers/api/v1/punches_controller_spec.rb:12]: You could test the case when the user is not authorized to access the punches or is unauthenticated.

File: spec/controllers/api/v1/punches_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    subject(:request) { get :index }
+    it { is_expected.to have_http_status(:ok) }
```
Comment (rhian-cs) [inline spec/controllers/api/v1/punches_controller_spec.rb:14]:
```suggestion
    subject(:request) { get :index }
    
    it { is_expected.to have_http_status(:ok) }
```
:nail_care:

**Review by rhian-cs — COMMENTED** [review]
Very nice!

File: spec/controllers/api/v1/punches_controller_spec.rb (hunk 2 above)
Comment (wenderjean) [reply spec/controllers/api/v1/punches_controller_spec.rb:8]: Don't we have a factory for the Punch model in order to use `create(:punch)`?

File: spec/controllers/api/v1/punches_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    it 'returns content type json' do
+      expect(request.content_type).to eq 'application/json; charset=utf-8'
+    end
+
+    it 'returns all punches for the current user' do
+      expect(JSON.parse(request.body)).to eq([
+        {
+          "created_at"=>"2022-04-26T12:53:51.248Z",
```
Comment (wenderjean) [inline spec/controllers/api/v1/punches_controller_spec.rb:23]: Can we use single quotes to keep the pattern?

**Review by wenderjean — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/punches_controller.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+module Api
+  module V1
+    class PunchesController < ApiController
+      # before_action :authenticate_user!
+
```
Comment (rhian-cs) [inline app/controllers/api/v1/punches_controller.rb:5]:
```suggestion
```
I don't think you need this commented code anymore.

**Review by rhian-cs — APPROVED** [review]
Looks great! :rocket:

File: spec/controllers/api/v1/punches_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
+require 'spec_helper'
+require  "#{Rails.root}/spec/support/controller_helpers.rb"
+
+describe Api::V1::PunchesController, :type => :controller do
+  let!(:user) { create(:user, :with_token) }
+  let!(:punches) { create(:punch, user_id: user.id) }
```
Comment (adbatista) [inline spec/controllers/api/v1/punches_controller_spec.rb:6]: where are you using this?

File: spec/controllers/api/v1/punches_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
+require 'spec_helper'
+require  "#{Rails.root}/spec/support/controller_helpers.rb"
+
+describe Api::V1::PunchesController, :type => :controller do
+  let!(:user) { create(:user, :with_token) }
```
Comment (adbatista) [inline spec/controllers/api/v1/punches_controller_spec.rb:5]: why the `let!` instead `let`

File: spec/support/controller_helpers.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
+RSpec.shared_examples 'an authenticated resource action' do |valid_header|
+  if valid_header
```
Comment (adbatista) [inline spec/support/controller_helpers.rb:2]: you don't use if in tests, create an shared example for the authenticated and not authenticated case

**Review by wenderjean — APPROVED** [review]

**Comment by adbatista** [comment]
just organize the commits and we should be fine on this :)

**Review by wenderjean — APPROVED** [review]
LGTM

**Pull request merged**